### ISSUE TITLE
CA1508 mis-fires in GetHashCode of generic class

### ISSUE THREAD

**Issue by AArnott** [dotnet/roslyn-analyzers#4579]
### Analyzer

**Diagnostic ID**: [CA1508](https://docs.microsoft.com/dotnet/fundamentals/code-analysis/quality-rules/ca1508): `Do not use ReferenceEquals with value types`

### Analyzer source

VS version: 16.9 Preview 3 (30815.8.main)
**SDK**: .NET SDK 5.0.100 via global.json

Other relevant versions:

```xml
    <PackageReference Include="Microsoft.CodeAnalysis.NetAnalyzers" Version="5.0.1" PrivateAssets="all" />
    <PackageReference Include="Microsoft.Net.Compilers.Toolset" Version="3.9.0-2.final" PrivateAssets="all" />
```

```cs
internal class DictionaryByValueEqualityComparer<TKey, TValue>
    where TKey : notnull
{
    public int GetHashCode(IReadOnlyDictionary<TKey, TValue> obj)
    {
        unchecked
        {
            int hash = obj.Count;

            foreach (KeyValuePair<TKey, TValue> entry in obj)
            {
                hash += entry.Key.GetHashCode();
                hash += entry.Value?.GetHashCode() ?? 0;
            }

            return hash;
        }
    }
}
```

### Describe the bug

The VS error list and msbuild.exe report that this line:

```cs
                hash += entry.Value?.GetHashCode() ?? 0;
```

> 1>D:\git\VS.RPC.Contracts\src\Microsoft.VisualStudio.RpcContracts\Utilities\DictionaryByValueEqualityComparer.cs(64,29,64,40): warning CA1508: 'entry.Value' is never 'null'. Remove or refactor the condition(s) to avoid dead code.
1>D:\git\VS.RPC.Contracts\src\Microsoft.VisualStudio.RpcContracts\Utilities\DictionaryByValueEqualityComparer.cs(64,29,64,55): warning CA1508: 'entry.Value?.GetHashCode()' is never 'null'. Remove or refactor the condition(s) to avoid dead code.

Yet in fact if I make the recommended changes the compiler emits another error.
The CA1508 diagnostic is wrong because in fact the type constraints (or lack thereof) on `TValue` do not preclude a nullable type argument from being used here.

**Comment by AArnott** [comment]
See also #4552. Per the advice in that bug, I updated the analyzer and compiler packages already to see if this was a dupe, but the warning remains.

**Comment by Evangelink** [comment]
I confirm that I can reproduce the issue with `5.0.1` but it seems to be fixed on `master` branch. What's weird is that the issue is still present on the pre-release `6.0.0-preview1.20609.1` and also on branch `release/6.0.1xx-preview1`.

@mavasani I don't know if there are some missing merge or if some were incorrect.

**Comment by mavasani** [comment]
Thanks a ton @Evangelink for all the analyses. Seems the auto-merge from `master` to `release/6.0.1xx-preview1` are blocked on merge conflicts in https://github.com/dotnet/roslyn-analyzers/pull/4527. I'll go ahead and get that PR fixed up.

Given your confirmation of the fix on master, I am going to close out this issue.
